### PR TITLE
[Components with more than 1 service] LPS-187067 Refactor components from frontend-js and frontend-theme modules

### PR DESCRIPTION
**Pull request by MrJohn1911**
Ticket: https://issues.liferay.com/browse/LPS-187067

Previous pull: https://github.com/liferay-core-infra/liferay-portal/pull/4971

### CODE REVIEW

**Comment by MrJohn1911** [comment]
ci:test:startup

**Comment by MrJohn1911** [comment]
ci:test:frontend

**Comment by catarineaguiar** [comment]
no regressions found

File: modules/apps/frontend-js/frontend-js-loader-modules-extender/src/main/java/com/liferay/frontend/js/loader/modules/extender/internal/servlet/JSResolveModulesNPMRegistryUpdatesListener.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import com.liferay.frontend.js.loader.modules.extender.npm.NPMRegistryUpdatesListener;
+
+import org.osgi.service.component.annotations.Component;
+
+/**
+ * @author Joao Victor Alves
+ */
+@Component(
+	configurationPid = "com.liferay.frontend.js.loader.modules.extender.internal.configuration.Details",
+	service = NPMRegistryUpdatesListener.class
```
Comment (rafaprax) [inline modules/apps/frontend-js/frontend-js-loader-modules-extender/src/main/java/com/liferay/frontend/js/loader/modules/extender/internal/servlet/JSResolveModulesNPMRegistryUpdatesListener.java:27]: Hey @MrJohn1911. you do not need to set the configurationPid here. This code does not depends on `com.liferay.frontend.js.loader.modules.extender.internal.configuration.Details` configuration. Please remove. Thx

File: modules/apps/frontend-js/frontend-js-loader-modules-extender/src/main/java/com/liferay/frontend/js/loader/modules/extender/internal/servlet/JSResolveModulesServlet.java [hunk 2]
```diff
@@ -159,7 +147,10 @@ private List<String> _getModuleNames(HttpServletRequest httpServletRequest)
 	@Reference
 	private BrowserModulesResolver _browserModulesResolver;
 
-	private volatile String _expectedPathInfo;
-	private volatile String _url;
+	@Reference(
+		target = "(component.name=com.liferay.frontend.js.loader.modules.extender.internal.servlet.JSResolveModulesNPMRegistryUpdatesListener)"
+	)
+	private NPMRegistryUpdatesListener
+		_jsResolveModulesNPMRegistryUpdatesListener;
```
Comment (rafaprax) [inline modules/apps/frontend-js/frontend-js-loader-modules-extender/src/main/java/com/liferay/frontend/js/loader/modules/extender/internal/servlet/JSResolveModulesServlet.java:154]: Hey @MrJohn1911, please see @dantewang's question here https://github.com/liferay-core-infra/liferay-portal/pull/4971#discussion_r1262254712

**Review by rafaprax — CHANGES_REQUESTED** [review]

**Comment by rafaprax** [comment]
Hey @MrJohn1911, please check if you addressed all @dantewang's questions and changes request. Thx

**Comment by MrJohn1911** [comment]
ci:test:sf

**Comment by MrJohn1911** [comment]
ci:test:relevant

**Comment by MrJohn1911** [comment]
ci:test:startup

**Comment by MrJohn1911** [comment]
ci:test:frontend

**Comment by catarineaguiar** [comment]
ci:test:stable

**Comment by catarineaguiar** [comment]
no regressions found

File: modules/apps/frontend-js/frontend-js-loader-modules-extender/src/main/java/com/liferay/frontend/js/loader/modules/extender/internal/servlet/JSResolveModulesServlet.java [hunk 3]
```diff
@@ -96,7 +83,7 @@ protected void service(
 			httpServletResponse.sendRedirect(
 				StringBundler.concat(
 					absolutePortalURLBuilder.forServlet(
-						getURL()
+						JSLoaderModulesUtil.getUrl()
```
Comment (rafaprax) [inline modules/apps/frontend-js/frontend-js-loader-modules-extender/src/main/java/com/liferay/frontend/js/loader/modules/extender/internal/servlet/JSResolveModulesServlet.java:86]: @MrJohn1911 please rename the method to `getURL` instead.

File: modules/apps/frontend-js/frontend-js-importmaps-extender/src/main/java/com/liferay/frontend/js/importmaps/extender/internal/servlet/taglib/util/JSImportMapsRegistryUtil.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+			return true;
+		}
+
+		return false;
+	}
+
+	public static JSImportMapsRegistration register(
+		String scope, JSONObject jsonObject) {
+
```
Comment (rafaprax) [inline modules/apps/frontend-js/frontend-js-importmaps-extender/src/main/java/com/liferay/frontend/js/importmaps/extender/internal/servlet/taglib/util/JSImportMapsRegistryUtil.java:48]: @MrJohn1911 please sort the parameters.

**Review by rafaprax — CHANGES_REQUESTED** [review]

File: modules/apps/frontend-js/frontend-js-importmaps-extender/src/main/java/com/liferay/frontend/js/importmaps/extender/internal/servlet/taglib/JSImportMapsExtenderTopHeadDynamicInclude.java [hunk 5]
```diff
@@ -65,8 +59,7 @@ public void include(
 		PrintWriter printWriter = httpServletResponse.getWriter();
 
 		if (_jsImportMapsConfiguration.enableImportMaps() &&
-			(!_globalImportMapJSONObjects.isEmpty() ||
-			 !_scopedImportMapJSONObjects.isEmpty())) {
+			JSImportMapsRegistryUtil.isBothNotEmptyImportMapsJSONObjects()) {
```
Comment (rafaprax) [inline modules/apps/frontend-js/frontend-js-importmaps-extender/src/main/java/com/liferay/frontend/js/importmaps/extender/internal/servlet/taglib/JSImportMapsExtenderTopHeadDynamicInclude.java:62]: @MrJohn1911, rename this to `isImportMapsJSONObjectsEmpty()` instead and apply the required changes. Thx

**Review by rafaprax — CHANGES_REQUESTED** [review]

**Comment by MrJohn1911** [comment]
ci:test:sf

**Comment by MrJohn1911** [comment]
ci:test:relevant

**Comment by MrJohn1911** [comment]
ci:test:startup

**Comment by MrJohn1911** [comment]
ci:test:frontend

**Comment by MrJohn1911** [comment]
ci:test:sf

**Comment by MrJohn1911** [comment]
ci:test:sf

**Comment by MrJohn1911** [comment]
Last push fixes sf problems

**Comment by Kvale1733** [comment]
ci:test:stable

**Comment by Kvale1733** [comment]
@MrJohn1911 No regressions found.

**Review by rafaprax — APPROVED** [review]

File: modules/apps/frontend-js/frontend-js-loader-modules-extender/src/main/java/com/liferay/frontend/js/loader/modules/extender/internal/servlet/taglib/JSLoaderTopHeadDynamicInclude.java [hunk 6]
```diff
@@ -131,7 +131,7 @@ private String _getResolvePath(HttpServletRequest httpServletRequest) {
 				httpServletRequest);
 
 		return absolutePortalURLBuilder.forServlet(
-			_jsResolveModulesServlet.getURL()
+			JSLoaderModulesUtil.getURL()
```
Comment (tinatian) [inline modules/apps/frontend-js/frontend-js-loader-modules-extender/src/main/java/com/liferay/frontend/js/loader/modules/extender/internal/servlet/taglib/JSLoaderTopHeadDynamicInclude.java:134]: @MrJohn1911 @rafaprax You may have a null value here.

Before your change, it is using _jsResolveModulesServlet to retrieve the url, and in the activate method of _jsResolveModulesServlet, the url is setup.

After your change, you won't be sure when the url is populated.

**Comment by tinatian** [comment]
@MrJohn1911 @rafaprax Please see my comments. Please also check your third commit, it seems it shares same issue as the second one.

You can send the first commit if it is not related to others.

Please resend.

Thanks.
Tina.

**Pull request closed without merging**